parse resource name before removing deleted secret

juanvallejo opened the pull request:
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1457602

Although unlinking deleted secrets from a serviceaccount is currently
supported, `oc secret unlink` fails to unlink a deleted secret if its
name is specified as `secrets/deleted-secret-name`.

This patch parses each secret's name, removing the `secrets/` segment
before appending it to a string set of removed secret names.

cc @openshift/cli-review

stevekuznetsov commented on pkg/oc/cli/secrets/options.go:
```diff
@@ -98,11 +99,22 @@ func (o SecretOptions) GetServiceAccount() (*kapi.ServiceAccount, error) {
 
+// parseResourceName receives either a resource name as either
+// <resource type> / <name> or <name> and returns only the resource <name>.
+func parseResourceName(name string) string {
```
This isn't meant to be generic, right? Maybe we could strip the `secret(s)/` prefix explicitly in that case. That would also be more clear in my opinion even if we are certain at this point that `name` has `secret/` as the prefix if it contains a slash. Otherwise, what happens if i `oc secret unlink whatever/steve`?

stevekuznetsov commented on test/cmd/secrets.sh:
```diff
@@ -85,6 +85,19 @@ os::cmd::expect_success 'oc secrets add deployer basicauth sshauth --for=pull'
+os::cmd::expect_success 'oc create secret generic deleted-secret'
+os::cmd::expect_success 'oc secrets link deployer deleted-secret'
+# confirm our soon-to-be-deleted secret has been linked
+os::cmd::expect_success 'oc get serviceaccounts/deployer -o yaml |grep -q deleted-secret'
```
Use `jsonpath` and validate that no secret is linked.


```
oc get serviceaccount deployer -o jsonpath='{.secrets[?(@.name=="deleted-secret")]}'
```

juanvallejo commented:
@stevekuznetsov thanks for the feedback, updated tests to use jsonpath; added check to ensure resource type is `secret/secrets` before "parsing" a user-specified secret name

stevekuznetsov commented on pkg/oc/cli/secrets/options.go:
```diff
@@ -98,11 +99,26 @@ func (o SecretOptions) GetServiceAccount() (*kapi.ServiceAccount, error) {
+		return name
+	}
+
+	if segs[0] == "secret" || segs[0] == "secrets" {
```
If you have a slash and this is not true, there is an error, right?

juanvallejo replied:
Right, the name will be returned as-is (e.g. `pods/my-secret`), and the user will end up seeing 
```
secret "pod/my-secret" not found
error: No valid secrets found or secrets not linked to service account
```

stevekuznetsov replied:
OK. I meant more forcefully -- the error _really_ is that a secret name cannot start with `pod/` not that `pod/my-secret` was not found

juanvallejo replied:
Sounds good, added a check for this in the `Validate` method

stevekuznetsov commented on test/cmd/secrets.sh:
```diff
@@ -85,6 +85,20 @@ os::cmd::expect_success 'oc secrets add deployer basicauth sshauth --for=pull'
+os::cmd::expect_success 'oc secrets link deployer deleted-secret'
+# confirm our soon-to-be-deleted secret has been linked
+os::cmd::expect_success_and_text "oc get serviceaccount deployer -o jsonpath='{.secrets[?(@.name==\"deleted-secret\")]}'" 'deleted\-secret'
+os::cmd::expect_success 'oc get serviceaccounts/deployer -o yaml |grep -q deleted-secret'
```
So when you unlink you should see no text here

stevekuznetsov replied:
In the JSONpath that is -- there should be no yaml grep in the final tests

juanvallejo replied:
Yeah, this is testing that the secret has linked. A few lines below from this, I test unlinking by calling `os::cmd::expect_success_and_not_text`

juanvallejo replied:
Also, forgot to remove the old `grep` check. Will remove that

stevekuznetsov commented:
/approve
/kind bug

juanvallejo commented:
/test extended_conformance_install_update

smarterclayton commented on pkg/oc/cli/secrets/options.go:
```diff
@@ -68,6 +69,16 @@ func (o SecretOptions) Validate() error {
 
+	// if any secret names are of the form <resource>/<name>,
+	// ensure <resource> is a secret.
+	for _, secretName := range o.SecretNames {
```
Why isn't this using standard resource builder?

juanvallejo replied:
>Why isn't this using standard resource builder?

The resource builder is [used in the `GetSecrets` method](https://github.com/openshift/origin/blob/master/pkg/oc/cli/secrets/options.go#L166). The code block here validates that a resource kind is `secret` if it is specified as `<kind>/<name>`, before getting to the builder, which ignores `NotFound` errors in order to allow  deleted secrets to still be unlinked. Without this check, if a user provides a command such as: `oc secret unlink sa_name pods/not_a_secret`, this line would create a secret with the name `pods/not_a_secret`.
